src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs

```diff
@@ -74,7 +75,91 @@ internal partial class IteratorRewriter : StateMachineRewriter
+            if (rewriter.EnsureAllSymbolsAndSignature())
+            {
+                return body;
+            }
+
+            return rewriter.Rewrite();
+        }
+
+        // returns true if any type we depend on is not present
+        protected bool EnsureAllSymbolsAndSignature()
+        {
+            bool hasErrors = false;
+
+            EnsureSpecialType(SpecialType.System_Int32, ref hasErrors);
+            EnsureSpecialMethod(SpecialMember.System_IDisposable__Dispose, ref hasErrors);
+
+            // IEnumerator<T>
+            EnsureSpecialPropertyAsMember(SpecialMember.System_Collections_Generic_IEnumerator_T__Current,
+                                            SpecialType.System_Collections_Generic_IEnumerator_T, ref hasErrors);
+
+            // IEnumerator
+            EnsureSpecialProperty(SpecialMember.System_Collections_IEnumerator__Current, ref hasErrors);
+            EnsureSpecialMethod(SpecialMember.System_Collections_IEnumerator__MoveNext, ref hasErrors);
+            EnsureSpecialMethod(SpecialMember.System_Collections_IEnumerator__Reset, ref hasErrors);
+
+            if (_isEnumerable)
+            {
+                // IEnumerable and IEnumerable<T>
+                EnsureSpecialMethod(SpecialMember.System_Collections_IEnumerable__GetEnumerator, ref hasErrors);
+                EnsureSpecialMethodAsMember(SpecialMember.System_Collections_Generic_IEnumerable_T__GetEnumerator,
+                                            SpecialType.System_Collections_Generic_IEnumerable_T, ref hasErrors);
+            }
+
+            return hasErrors;
+        }
+
+        private void EnsureSpecialProperty(SpecialMember type, ref bool hasErrors)
+        {
+            EnsureAvailableType(() => F.SpecialProperty(type).GetMethod, ref hasErrors);
+        }
+
+        private void EnsureSpecialMethod(SpecialMember type, ref bool hasErrors)
+        {
+            EnsureAvailableType(() => F.SpecialMethod(type), ref hasErrors);
+        }
+
+        private void EnsureSpecialType(SpecialType type, ref bool hasErrors)
+        {
+            EnsureAvailableType(() => F.SpecialType(type), ref hasErrors);
+        }
+
+        private void EnsureSpecialPropertyAsMember(SpecialMember property, SpecialType type, ref bool hasErrors)
+        {
+            EnsureAvailableType(() =>
+            {
+                var sType = F.SpecialType(type).Construct(_elementType);
+                return F.SpecialProperty(property).GetMethod.AsMember(sType);
+            }, ref hasErrors);
+        }
+
+        private void EnsureSpecialMethodAsMember(SpecialMember method, SpecialType type, ref bool hasErrors)
+        {
+            EnsureAvailableType(() =>
+            {
+                var sType = F.SpecialType(type).Construct(_elementType);
+                return F.SpecialMethod(method).AsMember(sType);
+            }, ref hasErrors);
+        }
+
+        private void EnsureAvailableType(Func<Symbol> getType, ref bool hasErrors)
```

Comment:
The name of the method is misleading, given how it is used. Also, I do not think we should be using lambdas and rely on BoundNodeFactory for the purpose of this checks. Let's talk offline in more details.